Wikipedia/Networking/Fetchers/SavedArticlesFetcher.m

```diff
@@ -27,59 +24,188 @@ @implementation SavedArticlesFetcher
 
 #pragma mark - Shared Access
 
-static SavedArticlesFetcher* _fetcher = nil;
+static SavedArticlesFetcher* _articleFetcher = nil;
 
 + (SavedArticlesFetcher*)sharedInstance {
-    return _fetcher;
+    NSParameterAssert([NSThread isMainThread]);
+    return _articleFetcher;
 }
 
-+ (void)setSharedInstance:(SavedArticlesFetcher*)fetcher {
-    _fetcher = fetcher;
++ (void)setSharedInstance:(SavedArticlesFetcher*)articleFetcher {
+    NSParameterAssert([NSThread isMainThread]);
+    _articleFetcher = articleFetcher;
 }
 
-- (instancetype)initWithDataStore:(MWKDataStore*)dataStore {
-    NSParameterAssert(dataStore != nil);
-
+- (instancetype)initWithArticleFetcher:(WMFArticleFetcher*)articleFetcher
+                       imageController:(WMFImageController*)imageController {
+    NSParameterAssert(articleFetcher);
     self = [super init];
     if (self) {
-        self.dataStore   = dataStore;
-        self.accessQueue = dispatch_queue_create("org.wikipedia.savedarticlesfetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
-        self.fetcher     = [[WMFArticleFetcher alloc] initWithDataStore:self.dataStore];
+        _accessQueue                       = dispatch_queue_create("org.wikipedia.savedarticlesarticleFetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
+        self.fetchOperationsByArticleTitle = [NSMutableDictionary new];
+        self.errorsByArticleTitle          = [NSMutableDictionary new];
+        self.articleFetcher                = articleFetcher;
+        self.imageController               = imageController;
     }
     return self;
 }
 
-- (void)fetchSavedPageList:(MWKSavedPageList*)savedPageList {
-    self.savedPageList = savedPageList;
+- (instancetype)initWithDataStore:(MWKDataStore*)dataStore {
+    return [self initWithArticleFetcher:[[WMFArticleFetcher alloc] initWithDataStore:dataStore]
+                        imageController:[WMFImageController sharedInstance]];
+}
+
+- (void)setSavedPageList:(MWKSavedPageList*)savedPageList {
+    // Using identity instead of equivalence since updates to an instance are more important than equivalent state
+    if (self.savedPageList == savedPageList) {
+        return;
+    }
+
+    [self.KVOControllerNonRetaining unobserve:self.savedPageList];
     [self cancelFetch];
 
+    _savedPageList = savedPageList;
+
+    [self fetchSavedPageList];
+}
+
+#pragma mark - Fetching
+
+- (void)fetchSavedPageList {
+    if (!self.savedPageList) {
+        return;
+    }
+
+    // build up initial state of current list
+    [self startFetching];
+
+    // observe subsequent changes
+    [self.KVOControllerNonRetaining observe:self.savedPageList
```

Comment:
Why wouldn't we just always observe the saved page list - why only start observing it in the fetch method.

The method seems to be conflating 2 discrete pieces of work:
- Observing the saved page list, and responding to changes
- Downloading articles for items in the saved page list

This conflation is demonstrated by the name of the method, `fetchSavedPageList`, which gives no indication that observation logic is going on in here.